pkg/decoders/utf16.go

```diff
@@ -31,25 +31,44 @@ func utf16ToUTF8(b []byte) ([]byte, error) {
 		return nil, err
 	}
 
-	buf := &bytes.Buffer{}
-	for i := 0; i < len(b); i += 2 {
-		r := rune(endianness.Uint16(b[i:]))
-		if utf8.ValidRune(r) {
-			buf.WriteRune(r)
+	var buf1, buf2 bytes.Buffer
+	var numLines1, numLines2 int
+	for i := 0; i < len(b)-1; i += 2 {
+		if r := rune(endianness.Uint16(b[i:])); utf8.ValidRune(r) {
+			buf1.WriteRune(r)
+			if r == '\n' {
+				numLines1++
+			}
+		}
+		// Guard against index out of bounds for the next check.
+		if i+1 >= len(b)-1 {
+			continue
+		}
+		// Same check but offset by one.
```

Comment:
Is the offset by one check in case the chunk was cut off in the middle of a character? If so, we should also invert the result of the endianness check.

Reply:
For the DLL case, the UTF-16 string was embedded within other binary data, so the guesser was guessing incorrectly.

I was thinking there were four ways to decode (endianness and even/odd starting index): (LE, even), (LE, odd), (BE, even), (BE, odd).

I'm not entirely convinced that the even/odd plays a role and it's really just BE vs LE though.

Reply:
What do you think of this https://github.com/trufflesecurity/trufflehog/compare/utf16-decoder-alt ?

It includes all valid BE or LE utf16 since both can exist together (like in the DLL). I'm could very well be wrong, but I think we just need to worry about BE or LE. The chunker will always cut off on an even byte, so we shouldn't get an off by one situation.

Reply:
Yeah I think that makes sense to do. I was leaning that way myself but wasn't sure about appending the two decoded strings together.. it should be fine, right?

Reply:
The main issue I see with it is that line numbers would be wrong, but that's out the window anyway when we're decoding partial files.